diff --git a/system/bluetooth/bluedroid/bluetooth.c b/system/bluetooth/bluedroid/bluetooth.c
--- a/system/bluetooth/bluedroid/bluetooth.c
+++ b/system/bluetooth/bluedroid/bluetooth.c
@@ -52,7 +52,11 @@ static int init_rfkill() {
     int fd;
     int sz;
     int id;
-    for (id = 0; ; id++) {
+    for (id = 1; ; id++) {
+        if (id > 1000)
+        {
+            return -1;
+        }
         snprintf(path, sizeof(path), "/sys/class/rfkill/rfkill%d/type", id);
         fd = open(path, O_RDONLY);
         if (fd < 0) {
