diff --git a/frameworks/base/core/java/android/hardware/Camera.java b/frameworks/base/core/java/android/hardware/Camera.java
--- a/frameworks/base/core/java/android/hardware/Camera.java
+++ b/frameworks/base/core/java/android/hardware/Camera.java
@@ -3842,4 +3842,6 @@ private boolean same(String s1, String s2) {
             return false;
         }
     };
+    // Allwinner platform - set video screen
+    public static native int setCameraScreen(int screen);
 }
diff --git a/packages/apps/Camera/src/com/android/camera/FocusOverlayManager.java b/packages/apps/Camera/src/com/android/camera/FocusOverlayManager.java
--- a/packages/apps/Camera/src/com/android/camera/FocusOverlayManager.java
+++ b/packages/apps/Camera/src/com/android/camera/FocusOverlayManager.java
@@ -405,8 +405,8 @@ public void onCameraReleased() {
 
     private void autoFocus() {
         Log.v(TAG, "Start autofocus.");
-        mListener.autoFocus();
-        mState = STATE_FOCUSING;
+        //mListener.autoFocus();
+        mState = STATE_SUCCESS;
         // Pause the face view because the driver will keep sending face
         // callbacks after the focus completes.
         if (mFaceView != null) mFaceView.pause();
diff --git a/system/core/include/system/camera.h b/system/core/include/system/camera.h
--- a/system/core/include/system/camera.h
+++ b/system/core/include/system/camera.h
@@ -160,6 +160,10 @@ enum {
      * Stop the face detection.
      */
     CAMERA_CMD_STOP_FACE_DETECTION = 7,
+    
+  // add command
+  CAMERA_CMD_SET_SCREEN_ID = 0xFF000000,
+  CAMERA_CMD_SET_CEDARX_RECORDER = 0xFF000001,
 
 #if defined(QCOM_ICS_COMPAT) && defined(QCOM_HARDWARE)
     CAMERA_CMD_HISTOGRAM_ON     = 8,
